### PR TITLE
Harden endowments

### PR DESCRIPTION
**Pull request by david0xd**
Fixes: https://github.com/MetaMask/snaps-monorepo/issues/1018

This PR will ensure that each endowment is passed to the `SES` `harden` function before exported.

Note: This is not final hardening work. This is one step closer to completing the epic. Half of the work is preparation after researching and identifying a nature of each endowment. Next steps might be continued and followed in the tickets: https://github.com/MetaMask/snaps-monorepo/issues/1018 and https://github.com/MetaMask/snaps-monorepo/issues/1017 (if needed).

## Attenuated endowments
Each (_existing_) attenuated endowment located under `/src/common/endowments` folder inside `snaps-execution-environments` repository is modified to handle exported functions or objects through `harden()`. All endowments of this type are tested for their security properties within other endowments in the `hardenedEndowments.ava.test.ts`.

## Common (default javascript) endowments
_All_ endowments* exported as `DEFAULT_ENDOWMENTS` from `snaps-utils` are now made as custom and all of them are passed to the `harden` function before final export.
All endowments are consolidated within `buildCommonEndowments` function that will provide factories for all of them, after harden is applied to each.

## Special endowments
There are two endowments considered as special: `ethereum` and `snap`. 
It is discovered that hardening of the `ethereum` endowment also known as `EIP-1193 Provider` cannot be done in the same place or within the `snaps-execution-environments` package as it is now, because `snaps-controllers` package is trying to update it in the process which is causing major failures. Further research is required on this, hence it is proposed that this work continue in the next task allocated in the same epic. Also, these endowments are very specific and encapsulated, hence, additional testing approaches might be required.

## Testing
Each endowment security is tested in `hardenedEndowments.ava.test.ts`. Only AVA test runner is running security related files and testing everything exposed to the functions from SES (harden).

There are four major scenarios applied for testing depending on the type of the endowment:
1. Adding new properties to the endowment should be impossible and result with an error.
2. Changing the prototype of the endowment should be impossible and result with an error.
3. Changing the __proto__ of the endowment should be impossible and result with an error.
4. Search for leak of the global `this` through object endowments or instances of the constructor functions.

There are three major endowment types identified during the research and testing. These are represented in the table below with test scenarios applied for each.

| Endowment type | Add new property (1) | Change prototype (2) | Change `__proto__` (3) | Search for global this |
| ---------------- | ---------------------- | --------------------- | ------------------- | ----------------------- |
| function              | Yes                                | Yes                                | Yes                            | No |
| constructor function (class)  | Yes         | Yes                                | Yes                           | Yes (instances) |
| object                 | Yes                                | No                                | Yes                            | Yes |

## Configuration changes

- `.c8rc.json` - Removed text report for coverage when running AVA tests (coverage report from AVA will not be reported in console independently anymore as it is redundant, merged coverage will be displayed at the end of the process).
- `nyc.config.js` - New configuration file added. This will enable coverage threshold check after coverage merge process is completed. So this will also enable failure and exit with code 1 when coverage thresholds are not satisfied. It is required because Jest cannot be responsible for the coverage threshold checks as it is only checking its own coverage during its run.
- `ava.config.js` - Removed `concurrency` settings since it's the best option for AVA to handle this automatically and by default it will distribute tests according to the availability of CPU cores. The `timeout` configurations is added and set to 30 seconds. This is required because AVA needs more time to start running, if the timeout is set too low or default, it can cause timeout failure during the test run on the CI pipeline. This does not means that it will use whole 30 seconds, but just how much is needed (part of it). Ava tests are running fine after the run has started. The only issue was the moment when it's loading tests which can take little bit longer and fail before it started.
- `jest.config.js` - Coverage threshold is removed so the Jest will not be checking it anymore (only merged coverage can be checked at the runtime). Coverage provider is returned to be `v8` since AVA is using the same and in order for merge process to work, it is needed to use the same coverage reporter.
- `jest.environments.js` - Because Jest has problem with running SES, `harden` function is mocked in the global environment with the function that will just return anything passed to it. This is the only way so far how it can work with `harden` involved, etc.
- `package.json` - `jest-it-up` is removed from usage, because if Jest configuration file is updated with the thresholds, it will then take the responsibility for threshold checks and fail on its own run (it will not use merged coverage, etc.). The `merge:coverage` command is updated so the merged coverage report will be displayed in the console after it completed testing process. Afterwards, a newly added script `update-coverage-thresholds.js` will run an update of the new coverage thresholds if required. 

## List of endowments and their _types_

| Endowment        | Type             |
| ---------------- | ---------------- |
| atob | function |
| btoa | function |
| BigInt | constructor |
| console | object |
| crypto | object |
| Date | constructor |
| Math | object |
| setTimeout | function |
| clearTimeout | function |
| SubtleCrypto | constructor |
| TextDecoder | constructor |
| TextEncoder | constructor |
| URL | constructor |
| WebAssembly | object |
| setInterval | function |
| clearInterval | function |
| Int8Array | constructor |
| Uint8Array | constructor |
| Uint8ClampedArray | constructor |
| Int16Array | constructor |
| Uint16Array | constructor |
| Int32Array | constructor |
| Uint32Array | constructor |
| Float32Array | constructor |
| Float64Array | constructor |
| BigInt64Array | constructor |
| BigUint64Array | constructor |
| DataView | constructor |
| ArrayBuffer | constructor |
| AbortController | constructor |
| AbortSignal | constructor |
| fetch | function |
| WebSocket | constructor |
| ethereum | object |
| snap | object |

Notes about types of endowments:
- `function` - All hardened in a simple way by applying `harden` function on them. Special cases such as `setTimeout`, `setInterval`, `fetch` ... are already handled in a wrapped way plus now additionally hardened with `harden`.
- `constructor` - All of these are hardened and tested.
- `object` - All of them are hardened and tested except the `ethereum` and `snap` endowment which is still in research process.

Additional notes:
Because of complexity and additional requirements `ethereum` and `snap` endowment should be hardened and tested in the other PR.

### CODE REVIEW

File: packages/snaps-execution-environments/src/common/endowments/abortController.ts
Comment (Mrtenz) [inline packages/snaps-execution-environments/src/common/endowments/abortController.ts:1]: A lot of these files are the same, except for the object being hardened. Couldn't you use an array of all the objects that need to be hardened, and to the hardening in a loop somehow? That would save a bunch of code duplication.

Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/abortController.ts:1]: Well, yes, somehow. Applying hardening is pretty much easy thing to do. But, it has to do things with how we provide and manage these endowments in the `index.ts` where it's expected to have a factory.

So, I tried to make this consistent with the other (previously attenuated) endowments that had some kind of modification (fetch, timeout, interval). Also, the tests would require hardening first where each endowment is tested on some atomic level, which does not involve `index.ts` and procedures for dependency injection implemented there. That's why each endowment is hardened on a factory level.

I could make most of these into one file with all of these factories and then export/test them like that when needed (still not sure if it's better or just more confusing). But, still there will be some as separate like `network`, `interval`, `timeout`, `crypto` which contain some heavy attenuation even before hardening. So, it was mostly because of consistency and trying to make clean understandable testing flow.

This can be an open question if there are any proposals or strong opinions. I'm happy to follow and implement whatever seems to be better if it doesn't add more functional or testing complexity!

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Alright, I added the following checks:
```javascript
instance.__proto__.__flag = 'not_secure';
subject.__proto__.__flag = 'not_secure';
```
With some TypeScript and ESlint hacks to make it run.

File: packages/snaps-execution-environments/src/common/endowments/abortController.ts
Comment (naugtur) [reply packages/snaps-execution-environments/src/common/endowments/abortController.ts:1]: Nothing forbids a factory factory

Comment (naugtur) [reply packages/snaps-execution-environments/src/common/endowments/abortController.ts:1]: And if that's where you go with it, you can have a single collection of endowments that you use for hardening and returning and also for testing them.

File: packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts:182]: I know I made you rearrange this test once already, but now that I look at it, it is testing whether the fields that were once exported are the same, so it's a perfect usecase of a snapshot.

You just need to put all the imported stuff in one object and snapshot it in a single test to acheive the same thing you did here.

File: packages/snaps-execution-environments/src/common/endowments/interval.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/endowments/interval.ts:24]: this makes freezing above redundant.

on the other hand, it's better to harden right away, so you should replace handle line with one of:
```js
    const handle = Object.freeze(Object.create(null));
```
or
```js
  const handle = harden({});
```

The first option being preferred as more explicit

File: packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts:326]: It's generally not possible for this test to ever fail the way timers are implemented with a Map keyed with object references, even if there was no protection against using other snaps' handlers. 

serialized and later deserialized, the handle is no longer the same object reference.

File: packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts
Comment (naugtur) [reply packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts]: Pushed the snapshot test

File: packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts
Comment (naugtur) [reply packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts]: The test could work if rewritten - see my comment

Comment (david0xd) [reply packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts]: I've removed this test as requested from the team and @ritave specifically here. 
I read your comment now but I also think that because we are prohibiting anything that is `unhardened endowment`, we don't have to do it in the tests. Perhaps it's alright to have it removed?

File: packages/snaps-execution-environments/src/common/endowments/interval.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/interval.ts]: I have added now the first option proposed with `const handle = Object.freeze(Object.create(null));`.
It is updated in both, timeout and interval endowments.

File: packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts
Comment (naugtur) [reply packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts]: This test was (incorrectly) testing a valid feature. IMHO it deserves keeping, rewritten to actually work. 
The idea is to pass the handle by setting it as a field of an unhardened endowment that both snaps get.

File: packages/snaps-execution-environments/src/common/endowments/abortController.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/abortController.ts]: To answer on this I've added endowment factory that builds consolidated endowment factories. There are still endowments that are attenuated in a custom way, like they were before. Others are hardened through the factory making process from a single collection. Only `TextDecoder` and `TextEncoder` are left as the separate modules since they're causing issues in the `snaps-controllers` because there is a problem with testing environments and availability of these endowments in `jest/jsdom`. At the end all of these common javascript endowments are provided by calling `buildCommonEndowments()` function where needed.

File: packages/snaps-execution-environments/src/common/endowments/security-utils/object-walker.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/endowments/security-utils/object-walker.ts:57]: this will only list enumerable properties. 

don't fix, I'll share a different implementation with you

File: packages/snaps-execution-environments/src/common/endowments/abortController.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/endowments/abortController.ts:18]: There's no need for this function to exist in more than one copy, at least as-is. 
Are you planning on making it dependent on the scope of its closure?

File: packages/snaps-execution-environments/src/common/endowments/abortController.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/endowments/abortController.ts:34]: To make it easier to reason about the signal object should be an instance of our dummy _AbortSignal class.
The handling of the signal itself needs a factory that binds an instance to an actual AbortSignal instance that is never accessible inside a snap but our _AbortSignal instance can be exchanged for the related AbortSignal instance outside of the snap.

That is done via a WeakMap where _AbortSignal instance is the key and AbortSignal instance is the value.
ALongside the factory, you also need a function that does the exchange. they both need to be kept private to our implementation only and share a scope in which the WeakMap lives.

File: packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts
Comment (ritave) [reply packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts]: This is getting into diminishing returns territory. Nuke it for now

File: packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts:323]: As agreed in the previous conversation the decision is still to remove the test: https://github.com/MetaMask/snaps-monorepo/pull/1058#pullrequestreview-1237336936
I will resolve this and related threads.

File: packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/BaseSnapExecutor.test.ts]: The final agreement is to remove the test as it is now on this PR. I will resolve this and related review threads on this PR.

File: packages/snaps-execution-environments/src/common/endowments/abortController.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/abortController.ts]: Thanks for the information provided.
But, as we agreed on slack this will be done in the next ticket. Because of the complexity for this case it is the best to keep this work separate.
The work will be followed in this ticket: https://github.com/MetaMask/snaps-monorepo/issues/1132

File: packages/snaps-execution-environments/src/common/endowments/abortController.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/abortController.ts]: As previously discussed we can separate and track this work within another ticket and PR.
https://github.com/MetaMask/snaps-monorepo/issues/1132
I will resolve this thread here now because this is now removed from this PR.

File: packages/snaps-execution-environments/jest.environment.js
Comment (naugtur) [inline packages/snaps-execution-environments/jest.environment.js:16]: Were Jest tests failing with the real `harden`?

File: packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts:22]: I thought the idea behind postponing Abort endowments was that we postpone adding them as endowments and not add them but postpone making them safe to add. 

Is this a decision that was made or just a result of circumstances?

File: packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ava.test.ts
Comment (naugtur) [inline packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ava.test.ts:21]: endowmentModules.ava.test.ts is asserting all that this file is asserting and more. This file seems redundant.

File: packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts:22]: `AbortController` and `AbortSignal` are both part of out default endowments which are exported and can be used for a long time.
I have no clear instructions on removing any of the default endowments, so I have to keep them all. 
Anyway, they'll get harden() at least in this iteration.
Here is the file with default endowments already exported or in use: https://github.com/MetaMask/snaps-monorepo/blob/main/packages/snaps-utils/src/default-endowments.ts

//cc @ritave

File: packages/snaps-execution-environments/jest.environment.js
Comment (david0xd) [reply packages/snaps-execution-environments/jest.environment.js:16]: Yes, the Jest was failing with `ReferenceError: harden is not defined` when that mock was not in the place.
Following your suggestion we decided to use only AVA for the tests that use `harden`.

File: packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts
Comment (naugtur) [reply packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts:22]: I'll approve the PR but encourage finishing this conversation before merge.

**Comment by naugtur** [comment]
when you rebase main you'll need to harden the new Date endowment too.

File: packages/snaps-execution-environments/src/common/endowments/security-utils/object-walker.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/security-utils/object-walker.ts]: Ok, this is resolved now by adding the object-walker from a lavamoat/walker package.

File: packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts:23]: We decided to leave these endowments as they were before exported in the defaults, then we will make a decision how and if we're going to keep them. We can address that work in the next ticket/PR: https://github.com/MetaMask/snaps-monorepo/issues/1132

Resolving this conversation for now.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (Mrtenz) [inline packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts:255]: If you use `@ts-expect-error`, you can remove the ESLint line above.

File: packages/snaps-execution-environments/jest.config.js
Comment (david0xd) [reply packages/snaps-execution-environments/jest.config.js:5]: I'm getting ESLint error if try the suggested code:
```
ESLint: Parsing error: Unexpected token ...
```
Our ESLint config does not support that maybe?

File: packages/snaps-execution-environments/package.json
Comment (david0xd) [reply packages/snaps-execution-environments/package.json:16]: Yes, removed now.

File: packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/commonEndowmentFactory.ts]: Changed.

File: packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts]: Updated.

File: packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts.md
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/endowmentModules.ava.test.ts.md]: It seems like it's not very useful, I've deleted it and added its extension to `.gitignore`.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts:28]: Added more documentation to this.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Changed.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Added more documentation to this.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (Mrtenz) [inline packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts:215]: This will disable the naming convention rule for the rest of the file. You can just put `/* eslint-disable @typescript-eslint/naming-convention */` once before both types, and `/* eslint-enable @typescript-eslint/naming-convention */` once after both types.

File: packages/snaps-execution-environments/update-coverage-thresholds.js
Comment (david0xd) [reply packages/snaps-execution-environments/update-coverage-thresholds.js]: Updated.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Updated.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: This is how Gal designed it to be used. It's taken from a demo: https://github.com/weizman/lavamoat-walker-demo-1/blob/main/index.js
It requires the assignment to work.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Updated as suggested.

File: packages/snaps-utils/src/post-process.ts
Comment (david0xd) [reply packages/snaps-utils/src/post-process.ts:121]: After I rebased two weeks ago, this started causing an error when I run the tests inside the `snaps-execution-environments`. It works in `snaps-utils`, but fails in the execution env.

Here is how it looks:
Screenshot 2023-01-23 at 19 37 36

So if I leave it as it was, tests will not work properly, which is annoying problem to have while developing :|

File: packages/snaps-execution-environments/src/common/endowments/index.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/index.ts:33]: Since I made a change to have factory for endowment factories, it started detecting this as an issue:
```
TS2345: Argument of type '() => unknown' is not assignable to parameter of type '() => EndowmentFactoryResult'.   Type 'unknown' is not assignable to type 'EndowmentFactoryResult'.
```

Because the starting value is `new Map<string, () => EndowmentFactoryResult>()` and `EndowmentFactoryResult` type (defined above) is not completely matching it.

Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/index.ts:33]: Let me know if you have a better solution to this. I can try to implement.

File: packages/snaps-utils/src/post-process.ts
Comment (Mrtenz) [reply packages/snaps-utils/src/post-process.ts:121]: Looks like there might be an issue with our `tsconfig.json` then? It should be set to ES2017, in which `matchAll` is not available.

Comment (david0xd) [reply packages/snaps-utils/src/post-process.ts:121]: I've tried adding `target: 'ES2017'` and `lib: ['ES2017']` to the:
`tsconfig.json`
`tsconfig.build.json`
`jest.config.js`

The error remains 🤔

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: I think it's not possible or I don't have a proper idea to rewrite this, since this is how it works now and it's the external library which I cannot change now. Only if we maybe request a change from the maintainers in the future. But, the Walker library is doing these things in a different way and is not very specific to our use case. 

It requires a callback which is passed here in the constructor. I do not see the other way to return `true` or `false` properly as the end result which I need for the test assertion.

File: packages/snaps-execution-environments/jest.config.js
Comment (david0xd) [reply packages/snaps-execution-environments/jest.config.js:5]: I did some research about using the error and using the suggested code. So, what I found is that I would need to change the ESLint config to have `ecmaVersion: 2018`. But, in our root directory there is a `tsconfig.packages.json` which has `"target": "ES2017"`, so that is a mismatch with the required ecma version for ESLint. I think I should leave it without changing these configurations for now.

File: packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts
Comment (weizman) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: '''
result = end === value;
return result;
'''

Should work too

Comment (weizman) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Is this what you asked for? @Mrtenz

Comment (Mrtenz) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Yeah, that seems more understandable.

Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/hardenedEndowments.ava.test.ts]: Yes it works, I've updated with the suggestion. Thanks!

File: packages/snaps-execution-environments/src/common/endowments/index.ts
Comment (david0xd) [reply packages/snaps-execution-environments/src/common/endowments/index.ts]: I've updated endowment factory type so this is no longer needed.

**Review by ritave — APPROVED** [review]

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**
